File: backend/src/db/schema/invoice_settings.ts
import { pgTable, text, timestamp, serial, boolean, doublePrecision, integer } from "drizzle-orm/pg-core";
import { companies } from "./companies";

export const invoiceSettings = pgTable("invoice_settings", {
    id: serial("id").primaryKey(),
    businessName: text("business_name").notNull(),
    layout: text("layout").notNull(),
    footerMessage: text("footer_message").notNull(),
    accentColor: text("accent_color").notNull(),
    fontFamily: text("font_family").notNull(),
    fontSizeMultiplier: doublePrecision("font_size_multiplier").notNull(),
    logoPath: text("logo_path"),
    logoLocalPath: text("logo_local_path"),
    logoBytes: text("logo_bytes"),
    thermalWidth: integer("thermal_width").default(80).notNull(),
    showLogo: boolean("show_logo").default(true).notNull(),
    showTaxBreakdown: boolean("show_tax_breakdown").default(true).notNull(),
    showAddress: boolean("show_address").default(true).notNull(),
    showCustomerDetails: boolean("show_customer_details").default(true).notNull(),
    showFooter: boolean("show_footer").default(true).notNull(),
    showBankDetails: boolean("show_bank_details").default(false).notNull(),
    showUpiQr: boolean("show_upi_qr").default(false).notNull(),
    bankName: text("bank_name"),
    bankAccountNo: text("bank_account_no"),
    bankIfsc: text("bank_ifsc"),
    upiId: text("upi_id"),
    businessPhone: text("business_phone"),
    businessAddress: text("business_address"),
    businessGstin: text("business_gstin"),
    customConfig: text("custom_config"),
    taxLabel: text("tax_label"),
    taxRate: doublePrecision("tax_rate"),
    termsAndConditions: text("terms_and_conditions"),
    authorizedSignatory: text("authorized_signatory"),
    invoiceCounter: integer("invoice_counter").default(0).notNull(),
    invoicePrefix: text("invoice_prefix").default("INV").notNull(),
    companyId: integer("company_id")
        .notNull()
        .references(() => companies.id, { onDelete: "cascade" }),
    updatedAt: timestamp("updated_at").defaultNow().notNull(),
});

export type InvoiceSettings = typeof invoiceSettings.$inferSelect;
export type NewInvoiceSettings = typeof invoiceSettings.$inferInsert;

File: ezo/lib/core/database/tables/invoice_settings.dart
import 'package:drift/drift.dart';

@DataClassName('InvoiceSettingEntity')
class InvoiceSettings extends Table {
  IntColumn get id => integer().autoIncrement()();
  TextColumn get layout => text().withLength(min: 1, max: 50)();
  TextColumn get footerMessage => text().withLength(max: 1000).withDefault(const Constant(''))();

  // Customization Options
  TextColumn get accentColor => text()
      .withLength(min: 1, max: 10)
      .withDefault(const Constant('#2A2D64'))();
  TextColumn get fontFamily => text()
      .withLength(min: 1, max: 50)
      .withDefault(const Constant('Roboto'))();
  RealColumn get fontSizeMultiplier =>
      real().withDefault(const Constant(1.0))();

  // Personalization
  TextColumn get logoPath => text().nullable()();
  TextColumn get logoLocalPath => text().nullable()();
  BlobColumn get logoBytes => blob().nullable()();
  
  // Thermal Options
  IntColumn get thermalWidth => integer().withDefault(const Constant(80))();

  // Bank & Payment Details
  TextColumn get bankName => text().nullable()();
  TextColumn get bankAccountNo => text().nullable()();
  TextColumn get bankIfsc => text().nullable()();
  TextColumn get upiId => text().nullable()();

  // Section Toggles
  BoolColumn get showLogo => boolean().withDefault(const Constant(true))();
  BoolColumn get showTaxBreakdown => boolean().withDefault(const Constant(true))();
  BoolColumn get showAddress => boolean().withDefault(const Constant(true))();
  BoolColumn get showCustomerDetails =>
      boolean().withDefault(const Constant(true))();
  BoolColumn get showFooter => boolean().withDefault(const Constant(true))();
  BoolColumn get showBankDetails =>
      boolean().withDefault(const Constant(false))();
  BoolColumn get showUpiQr =>
      boolean().withDefault(const Constant(false))();

  // Multi-tenant Link
  IntColumn get companyId => integer().nullable()();

  // Custom Template Configuration (JSON)
  TextColumn get customConfig => text().nullable()();

  // Tax configuration
  TextColumn get taxLabel => text().nullable()();
  RealColumn get taxRate => real().nullable()();

  // Footer / legal
  TextColumn get termsAndConditions => text().nullable()();
  TextColumn get authorizedSignatory => text().nullable()();

  // Invoice numbering
  IntColumn get invoiceCounter => integer().withDefault(const Constant(0))();
  TextColumn get invoicePrefix =>
      text().withDefault(const Constant('INV'))();
  TextColumn get deviceCode =>
      text().withDefault(const Constant('A'))();

  DateTimeColumn get updatedAt => dateTime().withDefault(currentDateAndTime)();
}
